File: src/main/java/me/pljr/pljrapispigot/PLJRApiSpigot.java
package me.pljr.pljrapispigot;

import lombok.Getter;
import me.pljr.pljrapispigot.config.Lang;
import me.pljr.pljrapispigot.config.Settings;
import me.pljr.pljrapispigot.config.SoundType;
import me.pljr.pljrapispigot.config.TitleType;
import me.pljr.pljrapispigot.database.DataSource;
import me.pljr.pljrapispigot.events.PLJRApiStartupEvent;
import me.pljr.pljrapispigot.listeners.BungeeListeners;
import me.pljr.pljrapispigot.listeners.PlayerQuitListener;
import me.pljr.pljrapispigot.managers.ConfigManager;
import me.pljr.pljrapispigot.managers.GUIManager;
import me.pljr.pljrapispigot.managers.QueryManager;
import me.pljr.pljrapispigot.utils.BStatsUtil;
import net.kyori.adventure.platform.bukkit.BukkitAudiences;
import net.milkbowl.vault.economy.Economy;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.plugin.PluginManager;
import org.bukkit.plugin.RegisteredServiceProvider;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.logging.Logger;

public final class PLJRApiSpigot extends JavaPlugin {
    private static final int BSTATS_ID = 10442;

    private static PLJRApiSpigot instance;
    public static Logger log;

    private DataSource dataSource;
    @Getter private Economy vaultEcon = null;
    @Getter private BukkitAudiences bukkitAudiences;
    @Getter private QueryManager queryManager;

    @Getter private GUIManager guiManager;

    private ConfigManager configManager;
    @Getter private Settings settings;

    public static PLJRApiSpigot get(){
        return instance;
    }

    @Override
    public void onEnable() {
        // Plugin startup logic
        instance = this;
        log = getLogger();
        BStatsUtil.addMetrics(this, BSTATS_ID);
        setupMiniMessage();
        setupConfig();
        setupDatabase();
        setupBungee();
        setupManagers();
        setupListeners();
        if (settings.isVault()) setupVault();
        getServer().getPluginManager().callEvent(new PLJRApiStartupEvent());
    }

    private void setupMiniMessage(){
        bukkitAudiences = BukkitAudiences.create(this);
    }

    private void setupVault(){
        if (!setupVaultEconomy()){
            log.severe(String.format(ChatColor.RED + "[%s] - Disabled due to no Vault dependency found!", getDescription().getName()));
            getServer().getPluginManager().disablePlugin(this);
        }
    }

    private boolean setupVaultEconomy(){
        if (getServer().getPluginManager().getPlugin("Vault") == null) {
            return false;
        }
        RegisteredServiceProvider<Economy> rsp = getServer().getServicesManager().getRegistration(Economy.class);
        if (rsp == null) {
            return false;
        }

        vaultEcon = rsp.getProvider();
        return vaultEcon != null;
    }

    private void setupDatabase(){
        dataSource = new DataSource(configManager);
        dataSource.initPool("PLJRAPI-Spigot-Pool");
        queryManager = new QueryManager(dataSource, this);
    }

    private void setupBungee(){
        getServer().getMessenger().registerOutgoingPluginChannel(this, "BungeeCord");
        getServer().getMessenger().registerIncomingPluginChannel(this, "pljrapi:chat", new BungeeListeners());
        getServer().getMessenger().registerOutgoingPluginChannel(this, "pljrapi:chat");
    }

    private void setupConfig(){
        saveDefaultConfig();
        configManager = new ConfigManager(this, "config.yml");
        settings = new Settings(configManager);
        Lang.load(new ConfigManager(this, "lang.yml"));
        SoundType.load(new ConfigManager(this, "sounds.yml"));
        TitleType.load(new ConfigManager(this, "titles.yml"));
    }

    private void setupManagers(){
        guiManager = new GUIManager(this);
    }

    private void setupListeners(){
        PluginManager pluginManager = Bukkit.getPluginManager();
        pluginManager.registerEvents(guiManager, this);
        pluginManager.registerEvents(new PlayerQuitListener(queryManager), this);
    }

    /**
     * Sends DataSource from provided configuration, if enabled.
     * PLJRApi's DataSource otherwise.
     *
     * @param config {@link DataSource} that will be checked for enabled MySQL.
     * @return {@link DataSource} from provided config, PLJRApi's otherwise.
     */
    public DataSource getDataSource(ConfigManager config){
        if (config.getBoolean("mysql.enabled")){
            return new DataSource(config);
        }
        return dataSource;
    }

    @Override
    public void onDisable() {
        instance = null;
    }
}

File: src/main/kotlin/me/pljr/pljrapispigot/PLJRApiSpigot.kt
package me.pljr.pljrapispigot

import me.pljr.pljrapispigot.bungeecord.BungeeListeners
import me.pljr.pljrapispigot.bungeecord.PLJRAPI_BUNGEE_CHANNEL
import me.pljr.pljrapispigot.config.ConfigManager
import me.pljr.pljrapispigot.config.configuration.Lang
import me.pljr.pljrapispigot.config.configuration.Settings
import me.pljr.pljrapispigot.config.configuration.SoundType
import me.pljr.pljrapispigot.config.configuration.TitleType
import me.pljr.pljrapispigot.database.DataSource
import me.pljr.pljrapispigot.events.PLJRApiSpigotStartupEvent
import me.pljr.pljrapispigot.gui.GUIListeners
import me.pljr.pljrapispigot.gui.GUIManager
import me.pljr.pljrapispigot.player.PlayerListeners
import me.pljr.pljrapispigot.player.PlayerQuery
import net.kyori.adventure.platform.bukkit.BukkitAudiences
import net.milkbowl.vault.economy.Economy
import org.bstats.bukkit.Metrics
import org.bukkit.Bukkit
import org.bukkit.plugin.java.JavaPlugin

class PLJRApiSpigot : JavaPlugin() {
    companion object {
        private const val BSTATS_ID = 10442
        lateinit var instance: PLJRApiSpigot

        lateinit var dataSource: DataSource
        var vaultEcon: Economy? = null
        lateinit var bukkitAudiences: BukkitAudiences

        lateinit var playerQuery: PlayerQuery

        lateinit var guiManager: GUIManager

        lateinit var configManager: ConfigManager

        /**
         * Sends DataSource from provided configuration, if enabled.
         * PLJRApi's DataSource otherwise.
         *
         * @param config [DataSource] that will be checked for enabled MySQL.
         * @return [DataSource] from provided config, PLJRApi's otherwise.
         */
        fun getDataSource(config: ConfigManager): DataSource {
            return if (config.getBoolean("mysql.enabled")) {
                DataSource(config)
            } else dataSource
        }
    }

    override fun onEnable() {
        instance = this
        Metrics(this, BSTATS_ID)
        setupAdventure()
        setupConfig()
        setupDatabase()
        setupBungee()
        setupManagers()
        setupListeners()
        if (Settings.VAULT) {
            setupVault()
        }
        server.pluginManager.callEvent(PLJRApiSpigotStartupEvent(this))
    }

    private fun setupAdventure() {
        bukkitAudiences = BukkitAudiences.create(this)
    }

    private fun setupConfig(){
        saveDefaultConfig()
        configManager = ConfigManager(this, "config.yml")
        Settings.load(configManager)
        Lang.load(ConfigManager(this, "lang.yml"))
        SoundType.load(ConfigManager(this, "sounds.yml"))
        TitleType.load(ConfigManager(this, "titles.yml"))
    }

    private fun setupDatabase() {
        dataSource = DataSource(configManager)
        dataSource.initPool("PLJRApi-Spigot-Pool")
        playerQuery = PlayerQuery(dataSource, this)
    }

    private fun setupBungee() {
        server.messenger.registerOutgoingPluginChannel(this, "BungeeCord")

        server.messenger.registerIncomingPluginChannel(this, PLJRAPI_BUNGEE_CHANNEL, BungeeListeners())
        server.messenger.registerOutgoingPluginChannel(this, PLJRAPI_BUNGEE_CHANNEL)
    }

    private fun setupManagers() {
        guiManager = GUIManager()
    }

    private fun setupListeners() {
        val pluginManager = Bukkit.getPluginManager()
        pluginManager.registerEvents(GUIListeners(this, guiManager), this)
        pluginManager.registerEvents(PlayerListeners(this, playerQuery), this)
    }

    private fun setupVault() {
        if (!setupVaultEconomy()) {
            logger.severe("Plugin disabled due to no Vault dependecy found!")
            server.pluginManager.disablePlugin(this)
        }
    }

    private fun setupVaultEconomy() : Boolean {
        if (server.pluginManager.getPlugin("Vault") != null) {
            val rsp = server.servicesManager.getRegistration(Economy::class.java)
            if (rsp != null) {
                vaultEcon = rsp.provider
                return true
            }
        }
        return false
    }
}